_studio/mfx_lib/shared/src/mfx_h264_enc_common_hw.cpp

```diff
@@ -725,6 +725,79 @@ namespace
         return mfxU32(384 * GetMaxMbps(par) / frameRate / GetMinCr(par.mfx.CodecLevel));
     }
 
+    mfxStatus CheckMaxFrameSize(MfxVideoParam & par, ENCODE_CAPS const & hwCaps)
+    {
+        mfxExtCodingOption2 *      extOpt2 = GetExtBuffer(par);
+        mfxExtCodingOption3 *      extOpt3 = GetExtBuffer(par);
+        bool changed = false;
+        bool unsupported = false;
+
+        if (extOpt2->MaxFrameSize == 0 && extOpt3->MaxFrameSizeI == 0 && extOpt3->MaxFrameSizeP == 0)
+            return MFX_ERR_NONE;
+
+        bool IsEnabledSwBrc = false;
+        IsEnabledSwBrc = bRateControlLA(par.mfx.RateControlMethod) || (IsOn(extOpt2->ExtBRC) && (par.mfx.RateControlMethod == MFX_RATECONTROL_CBR || par.mfx.RateControlMethod == MFX_RATECONTROL_VBR));
+
+        if ((par.mfx.RateControlMethod == MFX_RATECONTROL_CBR && // max fram size suported only for VBR based methods
+            par.mfx.RateControlMethod == MFX_RATECONTROL_CQP) ||
+            (hwCaps.UserMaxFrameSizeSupport == 0 && !IsEnabledSwBrc))
+        {
+            if (extOpt2->MaxFrameSize != 0 || extOpt3->MaxFrameSizeI != 0 || extOpt3->MaxFrameSizeP != 0)
+                changed = true;
+            extOpt2->MaxFrameSize = 0;
+            extOpt3->MaxFrameSizeI = 0;
+            extOpt3->MaxFrameSizeP = 0;
+        }
+        else
+        {
+            if (par.calcParam.cqpHrdMode == 0 && par.calcParam.targetKbps != 0 &&
+                par.mfx.FrameInfo.FrameRateExtN != 0 && par.mfx.FrameInfo.FrameRateExtD != 0)
+            {
+                mfxF64 frameRate = mfxF64(par.mfx.FrameInfo.FrameRateExtN) / par.mfx.FrameInfo.FrameRateExtD;
+                mfxU32 avgFrameSizeInBytes = mfxU32(par.calcParam.targetKbps * 1000 / frameRate / 8);
```

Comment:
Better mfxU32(par.calcParam.targetKbps / frameRate * (1000 / 8))
to avoid possible overflow of (par.calcParam.targetKbps * 1000) and optimize.